### ISSUE TITLE
Backward pass with sparse parameters results in error "Sparse division requires a scalar or zero-dim dense tensor divisor"

### ISSUE THREAD

**Issue by zabboud** [pytorch/pytorch#105304]
### 🐛 Describe the bug

I'm working on a simple model where the parameters are sparse and the inputs are dense. Let's take a simple Regression network, if we define the parameters as Sparse Tensors as follows: 

```python
class SparseLinear(nn.Module):
  def __init__(self, in_features, out_features, rand=True, rate=0.1):
    super().__init__()

    w = torch.FloatTensor(in_features, out_features)
    b = torch.zeros(out_features)
    if rand:
        w = torch.empty(in_features, out_features)
        nn.init.sparse_(w, sparsity=rate, std=0.01)
        w = w.to_sparse()

    self.weight = nn.Parameter(data=w, requires_grad=True)
    self.bias = nn.Parameter(data=b, requires_grad=True)

  def forward(self, x):
    return torch.sparse.addmm(self.bias, self.weight.T, x.T)
```

If we instantiate a network with a single input layer (90), a single hidden layer (100 features), an output layer of size (1) (YearPredictionMSD dataset is used here) 

```python
class Regression(nn.Module):
    def __init__(self, input_features:int, hidden_features:int, sparse:dict, output_features=1):
        super().__init__() 
        self.nonlinearity = lambda x: F.relu(x, inplace=True)
 
        self.fc1 = SparseLinear(input_features, hidden_features, rate=sparse['rate'], rand=sparse['random'])
        self.fc2 = SparseLinear(hidden_features, output_features, rate=sparse['rate'], rand=sparse['random'])

    def forward(self, x):
        x = self.nonlinearity(self.fc1(x))
        x = self.fc2(x)
        return x
```

Using `nn.MSE()` as the loss, an exception occurs at the backward pass 

`Exception has occurred: RuntimeError
Sparse division requires a scalar or zero-dim dense tensor divisor (got shape [1, 1] for divisor)`

Not really sure what this is referring to. There was an[ issue](https://discuss.pytorch.org/t/bug-in-backprop-with-sparse-tensors/175835) submitted where the target tensor is sparse (in my case, inputs and targets are dense, weights are sparse), where the NLL loss was used in that case, and a suggestion was made to re-write the loss with its backward pass to avoid a division by 0. 

However, in my case with the MSE loss, this is not an issue (also I've tested it with a custom loss and extended the `torch.autograd.Function` with the backward pass, but the same error occurs still. 

Any possibility of a work around to have the loss propagated through sparse weights? Should I be using hooks? 

Any suggestion would be helpful!

### Versions

PyTorch version: 2.0.1
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: CentOS Stream 9 (x86_64)
GCC version: (GCC) 11.3.1 20221121 (Red Hat 11.3.1-4)
Clang version: 15.0.7 (Red Hat 15.0.7-2.el9)
CMake version: version 3.20.2
Libc version: glibc-2.34

Python version: 3.8.12 | packaged by conda-forge | (default, Jan 30 2022, 23:42:07)  [GCC 9.4.0] (64-bit runtime)
Python platform: Linux-5.14.0-282.el9.x86_64-x86_64-with-glibc2.10
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 3090
Nvidia driver version: 525.78.01
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Address sizes:                   39 bits physical, 48 bits virtual
Byte Order:                      Little Endian
CPU(s):                          16
On-line CPU(s) list:             0-15
Vendor ID:                       GenuineIntel
Model name:                      11th Gen Intel(R) Core(TM) i7-11700K @ 3.60GHz
CPU family:                      6
Model:                           167
Thread(s) per core:              2
Core(s) per socket:              8
Socket(s):                       1
Stepping:                        1
CPU max MHz:                     5000.0000
CPU min MHz:                     800.0000
BogoMIPS:                        7200.00
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb invpcid_single ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid mpx avx512f avx512dq rdseed adx smap avx512ifma clflushopt intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp hwp_pkg_req avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq rdpid fsrm md_clear flush_l1d arch_capabilities
Virtualization:                  VT-x
L1d cache:                       384 KiB (8 instances)
L1i cache:                       256 KiB (8 instances)
L2 cache:                        4 MiB (8 instances)
L3 cache:                        16 MiB (1 instance)
NUMA node(s):                    1
NUMA node0 CPU(s):               0-15
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Retbleed:          Mitigation; Enhanced IBRS
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Enhanced IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequence
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected

Versions of relevant libraries:
[pip3] numpy==1.24.3
[pip3] torch==2.0.1
[pip3] torchaudio==2.0.2
[pip3] torchvision==0.15.2
[pip3] triton==2.0.0
[conda] blas                      1.0                         mkl  
[conda] ffmpeg                    4.3                  hf484d3e_0    pytorch
[conda] libblas                   3.9.0            12_linux64_mkl    conda-forge
[conda] libcblas                  3.9.0            12_linux64_mkl    conda-forge
[conda] liblapack                 3.9.0            12_linux64_mkl    conda-forge
[conda] liblapacke                3.9.0            12_linux64_mkl    conda-forge
[conda] mkl                       2021.4.0           h06a4308_640  
[conda] mkl-service               2.4.0            py38h95df7f1_0    conda-forge
[conda] mkl_fft                   1.3.1            py38h8666266_1    conda-forge
[conda] mkl_random                1.2.2            py38h1abd341_0    conda-forge
[conda] numpy                     1.24.3           py38h14f4228_0  
[conda] numpy-base                1.24.3           py38h31eccc5_0  
[conda] pytorch                   2.0.1           py3.8_cuda11.8_cudnn8.7.0_0    pytorch
[conda] pytorch-cuda              11.8                 h7e8668a_5    pytorch
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] torchaudio                2.0.2                py38_cu118    pytorch
[conda] torchtriton               2.0.0                      py38    pytorch
[conda] torchvision               0.15.2               py38_cu118    pytorch

**Comment by mikaylagawarecki** [comment]
It seems like there might be some support missing for sparse tensors. Could you provide a small repro that yields the erro please?

For the second error message with `foreach_mul` are you using the optimizer with `foreach=True`? Perhaps try `foreach=False` to workaround

**Comment by zabboud** [comment]
@mikaylagawarecki Thanks for your prompt reply. 

I put together a full working code with a toy dataset on google colab, please find it in this [link](https://colab.research.google.com/drive/1kEavptb4ewhOGZd7szFOy-7tyTb4RQNr?usp=sharing). I tested it with `regularize=True` it reproduces the first error. However, I'm unable to reproduce the second error with `foreach=True` with this dummy dataset. 

You can toggle regularization on and off using the `regularize=True` variable, which is under the "Training" section under `###TRAINING PARAMETERS`

I tested with `foreach=False` and yes, the error no longer occurs, however with the larger dataset (the YearPredictionMSD) the loss explodes to infinity pretty quick. Note that I'm using SGD as adam does not support sparse gradients and SparseAdam does not support sparse parameters, only dense parameters with sparse gradients.
